### DISCUSSION TITLE
Storybook build not working after babel-core patch bump

### DISCUSSION THREAD

**Discussion by gabrielbahniuk** [babel/babel#14439] [Q&A]
Hi there,

in my React project I'm using `@babel/core` version `7.17.7`. I just wanted to perform the patch update to `7.17.8`,however I get the following error when running my `storybook` build with the command `build-storybook -c storybook -o .storybook-static`

```
 'loose' mode configuration must be the same for @babel/plugin-proposal-class-properties, @babel/plugin-proposal-private-methods and @babel/plugin-proposal-private-property-in-object (when they are enabled).
```

To "solve" this, I had to add `@babel/plugin-proposal-class-properties` to my devDependencies in package.json and change babel configuration, adding `@babel/plugin-proposal-class-properties` as plugin with an object `{loose: true}`.

Maybe I am missing some detail, but I don't understand how a patch version broke my build.

I'm using the following addon versions in storybook:

```
"@storybook/addon-actions": "6.4.20",
"@storybook/addon-controls": "6.4.20",
"@storybook/addon-docs": "6.4.20",
"@storybook/addon-viewport": "6.4.20",
```

Thanks in advance for any hint that can lead to a solution.

**Comment by nicolo-ribaudo** [comment]
Can you share your previous config?

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo thanks for the quick answer!

I didn't change any config, so there's no previous, but I can share the current one:

```js
module.exports = api => {
    const plugins = [
        '@babel/plugin-transform-runtime',
        '@babel/plugin-proposal-class-properties',
        '@babel/plugin-proposal-optional-chaining'
    ];

    let presets = ['@babel/preset-react', '@babel/typescript'];

    if (process.env.NODE_ENV === 'test') {
        presets = ['next/babel'];
    }

    return {
        presets,
        plugins
    };
};
```

**Reply by gabrielbahniuk** [reply]
### Important

The current setup is that this babel configuration comes from a "central project". In my project I just `require` the file above from the other project as a npm package.

That said, this "central project" which takes care of the libs has `@babel/core` version `7.12.10`. Not sure if this version is incompatible with `7.17.8`.

**Reply by nicolo-ribaudo** [reply]
Are you sure you don't have another config?

That error should only happen when two of those plugins are enabled with different `loose` modes, but:
1. you don't have `loose: true` anywhere
2. The only preset you use doesn't have `loose: true` anywhere (https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts)


Could you try running this command to see your full config?
```
BABEL_SHOW_CONFIG_FOR=./path/to/js/file/you/are/compiling.ts build-storybook -c storybook -o .storybook-static
```

**Reply by JLHwung** [reply]
Can you print the config? https://babeljs.io/docs/en/configuration#print-effective-configs

The error is thrown when you have loaded '@babel/plugin-proposal-class-properties' more than once with different `"loose"` settings, but the config you have posted only include one '@babel/plugin-proposal-class-properties'. So I guess there are more configs.

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo I can't print the config bc I run into original posted error when I try `BABEL_SHOW_CONFIG_FOR=./storybook/main.js build-storybook -c storybook -o .storybook-static`. After reading the link in the docs by @JLHwung I also didn't get to which file I should point to. I assume it's not to my main entry file right? Because my app build works fine, just storybook has this issue.

Btw I just found within `storybook` folder I have `main.js` and `preview.tsx`, not sure if that helps somehow:

main.js:
```js
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');

const customWebpackConfig = async config => {
    /* eslint-disable no-param-reassign */
    config.resolve.plugins = [
        ...(config.resolve.plugins || []),
        new TsconfigPathsPlugin({
            extensions: config.resolve.extensions
        })
    ];
    /* eslint-enable no-param-reassign */

    return config;
};

module.exports = {
    stories: ['../src/**/*.stories.@(js|jsx|ts|tsx)'],
    addons: ['@storybook/addon-actions', '@storybook/addon-viewport', '@storybook/addon-controls'],
    webpackFinal: customWebpackConfig
};
```

preview.tsx:

```js
[...]
// imports here
[...]


setupI18n(en);

const StorybookLayout = ({ children }) => <main>{children}</main>;

addDecorator(story => (
    <AssetPrefixProvider assetPrefix="/public">
        <CurrencyProvider currency={CURRENCY}>
            <MuiThemeProvider theme={theme}>
                <StorybookLayout>{story()}</StorybookLayout>
            </MuiThemeProvider>
        </CurrencyProvider>
    </AssetPrefixProvider>
));

addDecorator(story => (
    <>
        <CssBaseline />
        {story()}
    </>
));

addDecorator(story => (
    <StylesProvider
        jss={create({
            plugins: [...customJssPreset().plugins]
        })}
    >
        {story()}
    </StylesProvider>
));
```

I also printed my config returned from `main.js` file, in case it's from any help:

```js
{
  config: {
    name: 'preview',
    mode: 'production',
    bail: true,
    devtool: 'cheap-module-source-map',
    entry: [
      'path/to/node_modules/@storybook/core-client/dist/esm/globals/polyfills.js',
      'path/to/node_modules/@storybook/core-client/dist/esm/globals/globals.js',
      'path/to/storybook-init-framework-entry.js',
      'path/to/@storybook/react/dist/esm/client/preview/config-generated-config-entry.js',
      'path/to/@storybook/addon-actions/dist/esm/preset/addDecorator.js-generated-config-entry.js',
      'path/to/@storybook/addon-actions/dist/esm/preset/addArgs.js-generated-config-entry.js',
      'path/to/storybook/preview.tsx-generated-config-entry.js',
      'path/to/generated-stories-entry.js'
    ],
    output: {
      path: './path/to/.storybook-static',
      filename: '[name].[contenthash:8].iframe.bundle.js',
      publicPath: ''
    },
    watchOptions: { ignored: /node_modules/ },
    plugins: [
      [FilterWarningsPlugin],
      [VirtualModulesPlugin],
      [HtmlWebpackPlugin],
      [DefinePlugin],
      [CaseSensitivePathsPlugin],
      [ProgressPlugin],
      [DocgenPlugin]
    ],
    module: { rules: [Array] },
    resolve: {
      extensions: [Array],
      modules: [Array],
      mainFields: [Array],
      alias: [Object],
      plugins: [Array]
    },
    resolveLoader: { plugins: [Array] },
    optimization: {
      splitChunks: [Object],
      runtimeChunk: true,
      sideEffects: true,
      usedExports: true,
      moduleIds: 'named',
      minimizer: [Array]
    },
    performance: { hints: 'warning' }
  }
}
```

Also I just tried to update `@babel/preset-react` and `@babel/runtime-corejs2` to newest versions, and it works fine, it's really only `@babel/core` >= 7.17.8 the problem. 

I didn't check all changes between `7.17.7` and `7.17.8` but it seems there's a breaking change!? Or maybe something was fixed and now I'm facing this issue.

**Reply by nicolo-ribaudo** [reply]
Can you try editing `@babel/helper-create-class-features-plugin/lib/features.is` in your `node_modules`, add `Error.stackTraceLimit = Infinity` before the line where that error is thrown, and post the full stack trace?

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo I changed to this in `node_modules/@babel/plugin-proposal-class-properties/node_modules/@babel/helper-create-class-features-plugin/lib/features.js`:

```js
  for (const [mask, name] of featuresSameLoose) {
    if (!hasFeature(file, mask)) continue;
    const loose = isLoose(file, mask);
    if (canIgnoreLoose(file, mask)) {
      continue;
    } else if (resolvedLoose === !loose) {
      Error.stackTraceLimit = Infinity; // here's what I added
      throw new Error("'loose' mode configuration must be the same for @babel/plugin-proposal-class-properties, " + "@babel/plugin-proposal-private-methods and " + "@babel/plugin-proposal-private-property-in-object (when they are enabled).");
    } else {
      resolvedLoose = loose;
      higherPriorityPluginName = name;
    }
```

Stacktrace: https://gist.github.com/gabrielbahniuk/f795c1f7055c3804ea85e962edf910ab

**Reply by nicolo-ribaudo** [reply]
Thank you! Can you now trying to run this command, to see the config of the file which is causing Babel to throw the error? (I got that path from your log, replace `my/home/dir/my-project` with the correct path!)

```
BABEL_SHOW_CONFIG_FOR=my/home/dir/my-project/generated-stories-entry.js build-storybook -c storybook -o .storybook-static
```

you might have to try a few times to get the result, because from that output log it looks like storybook is calling Babel in parallel multiple times and thus the build process could stop due to the same error on a different file before trying to compile `my/home/dir/my-project/generated-stories-entry.js`.

**Reply by nicolo-ribaudo** [reply]
Can you also checks what happens if you remove `loose: true` in these two lines? https://github.com/storybookjs/storybook/blob/c35ea5aa79312b8346f0f3cd15a1545db676b105/lib/core-common/src/utils/babel.ts#L22-L23

They should be somewhere in your `node_modules` in a `@storybook/core-common`. If that Babel configuration is loaded and merged with your main config (I don't understand enough storybook's internals to tell what's happening by just looking at their code), then you are forced to use `@babel/plugin-proposal-class-properties` with loose mode also when enabling it in your "personal" Babel config. If that's the problem, I admit I don't know how it's possible that it worked with the previous `@babel/core` version :sweat_smile: I would expected it to already fail (because the two configs, enabling the same plugin with loose mode and without loose mode, cannot be merged together in a "sensible" way).

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo thanks for the help!

Somehow I don't have this `generated-stories-entry.js` file - even though it appears in my logs, maybe it's generated and already deleted due to failure!? I tried multiple times but no success here.

Regarding your 2nd comment, removing `loose: true` object or replacing `true` by `false` in `node_modules/@storybook/core-common/dist/cjs/utils/babel.js` 
 (https://github.com/storybookjs/storybook/blob/c35ea5aa79312b8346f0f3cd15a1545db676b105/lib/core-common/src/utils/babel.ts#L23) made my build-storybook work, although with repeated warnings:

```
Though the "loose" option was set to "true" in your @babel/preset-env config, it will not be used for @babel/plugin-proposal-private-property-in-object since the "loose" mode option was set to "false" for @babel/plugin-proposal-private-methods.
The "loose" option must be the same for @babel/plugin-proposal-class-properties, @babel/plugin-proposal-private-methods and @babel/plugin-proposal-private-property-in-object (when they are enabled): you can silence this warning by explicitly adding
        ["@babel/plugin-proposal-private-property-in-object", { "loose": false }]
```

Now I know neither how it was working in version `7.17.7` and no more in babel/core 7.17.8+ nor where's the root cause of this problem.

**Reply by nicolo-ribaudo** [reply]
Probably `generated-stories-entry.js` is a "virtual" file generated to storybook and passed to webpack/babel, without ever being a real file on the filesystem.

I investigated a bit more, and it the source of the problem might be that Storybook enables the `class-properties` plugin in loose mode, wihle Next.js enables it without loose mode (https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts).

You can probably make it work by passing `{ "class-properties": { "loose": true } }` as an option to `next/babel`, so that it uses loose mode too.

It's still a mystery to me how it could work before your upgrades, since that error has been there for long time.

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo using the following has no effects as well:

```js
module.exports = api => {
    const plugins = [
        '@babel/plugin-transform-runtime',
        '@babel/plugin-proposal-class-properties',
        '@babel/plugin-proposal-optional-chaining'
    ];

    let presets = ['@babel/preset-react', '@babel/typescript'];

    if (process.env.NODE_ENV === 'test') {
        presets = [
            'next/babel',
            {
                'class-properties': { loose: true }
            }
        ];
    }

    return {
        presets,
        plugins
    };
};
```

However it works (with warnings) if I add the following:
```
    const plugins = [
        '@babel/plugin-transform-runtime',
        ['@babel/plugin-proposal-class-properties', { loose: true }],
        '@babel/plugin-proposal-optional-chaining'
    ];
```
but it's a mystery for me how it was working and also why it stopped working after 1 patch update - i still didn't check the diff between babel/core 7.17.7 and 7.17.8 but will try to find something there.

In any case, a big & honest thank you @nicolo-ribaudo  for all the support you provided, it helped me to find a way to make this work with only 1 line change, even though I still didn't understand the reason.

**Reply by nicolo-ribaudo** [reply]
If this works, it's slightly better (because you avoid enabling a third copy of the class properties plugin):
```js
module.exports = api => {
    const plugins = [
        '@babel/plugin-transform-runtime',
        // NOTE: I removed this
        // '@babel/plugin-proposal-class-properties'
        '@babel/plugin-proposal-optional-chaining'
    ];

    let presets = ['@babel/preset-react', '@babel/typescript'];

    if (process.env.NODE_ENV === 'test') {
        presets = [
            'next/babel',
            {
                'class-properties': { loose: true }
            }
        ];
    }

    return {
        presets,
        plugins
    };
};
```

Btw, `@babel/plugin-proposal-optional-chaining` might not be needed too because I believe that `next/babel` already enables it.

**Reply by gabrielbahniuk** [reply]
@nicolo-ribaudo this works fine for both builds (app + storybook):

```js
   const plugins = [
        '@babel/plugin-transform-runtime'
        //  '@babel/plugin-proposal-class-properties',
        // '@babel/plugin-proposal-optional-chaining'
    ];
```

with or without `class-properties` in `next/babel`. btw the problem was that I forgot an additional bracket, that's why `class-properties` didn't work before.

So if I understand correctly, `next/babel` already enables loose by default and i don't need to add this to the config, only commenting both plugins above seems to work, if I didn't break anything else

**Reply by nicolo-ribaudo** [reply]
`next/babel` does not enable loose by default, that's why you should need to also pass `'class-properties': { loose: true }` to its options!

**Reply by gabrielbahniuk** [reply]
you are right, I mixed things up between storybook and next.

**Comment by gabrielbahniuk** [comment]
I did not see which peace of code solved the issue, but babel/core@7.17.10 seems not to have it, I just tested and build storybook works fine again, without additional changes from my side.

Maybe not that relevant, but i just wanted to mention it
